File: legacy/services/seo-service/app/keyword_clustering.py
"""
Real Keyword Clustering — Section 4.1.1.
Groups keywords into topical clusters using Jaccard similarity on word sets.
This is a genuine, deterministic algorithm — not as semantically rich as an
embedding-based approach (which the blueprint's production version uses,
Section 7.2.1 step 4), but real, free, and doesn't require any paid API or
model. Good enough to group "office chair" with "ergonomic office chair"
and "best office chair" while keeping "dining table" separate.
"""
import re
from typing import List, Dict


STOPWORDS = {"a", "an", "the", "for", "with", "of", "in", "on", "to", "and", "best", "top"}


def _tokenize(keyword: str) -> set:
    words = re.findall(r"[a-z0-9]+", keyword.lower())
    return {w for w in words if w not in STOPWORDS}


def _jaccard_similarity(a: set, b: set) -> float:
    if not a or not b:
        return 0.0
    intersection = len(a & b)
    union = len(a | b)
    return intersection / union if union else 0.0


def cluster_keywords(keywords: List[str], similarity_threshold: float = 0.3) -> List[Dict]:
    """
    Groups keywords sharing enough overlapping significant words into the
    same cluster. Returns a list of {clusterName, keywords} dicts.
    """
    token_sets = {kw: _tokenize(kw) for kw in keywords}
    assigned = set()
    clusters = []

    for kw in keywords:
        if kw in assigned:
            continue
        cluster = [kw]
        assigned.add(kw)
        for other in keywords:
            if other in assigned:
                continue
            if _jaccard_similarity(token_sets[kw], token_sets[other]) >= similarity_threshold:
                cluster.append(other)
                assigned.add(other)

        # Name the cluster after the most common significant word across its members
        word_counts: Dict[str, int] = {}
        for member in cluster:
            for word in token_sets[member]:
                word_counts[word] = word_counts.get(word, 0) + 1
        cluster_name = max(word_counts, key=word_counts.get).title() if word_counts else cluster[0]

        clusters.append({"clusterName": cluster_name, "keywords": cluster})

    return clusters

File: packages/logic/src/engines/keyword-clustering.ts
/**
 * Real keyword clustering (SEO extras — restored from legacy/services/seo-service).
 *
 * Groups keywords into topical clusters using Jaccard similarity on their significant word sets.
 * Not as semantically rich as an embedding-based approach, but a genuine, deterministic algorithm
 * — real, free, and needs no third-party API or model. Good enough to group "office chair" with
 * "ergonomic office chair" and "best office chair" while keeping "dining table" separate.
 *
 * Pure and synchronous on purpose: unlike the site-audit crawler (real I/O, so it runs as a
 * background worker job), this only touches an in-memory list of strings, so it runs directly in
 * the API request — no job/poll round trip needed for something that completes in milliseconds.
 */

const STOPWORDS = new Set([
  'a', 'an', 'the', 'for', 'with', 'of', 'in', 'on', 'to', 'and', 'best', 'top',
]);

function tokenize(keyword: string): Set<string> {
  const words = keyword.toLowerCase().match(/[a-z0-9]+/g) ?? [];
  return new Set(words.filter((w) => !STOPWORDS.has(w)));
}

function jaccardSimilarity(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 || b.size === 0) return 0;
  let intersection = 0;
  for (const w of a) if (b.has(w)) intersection++;
  const union = a.size + b.size - intersection;
  return union === 0 ? 0 : intersection / union;
}

export interface KeywordCluster {
  clusterName: string;
  keywords: string[];
}

/** Groups keywords sharing enough overlapping significant words into the same cluster. */
export function clusterKeywords(
  keywords: string[],
  similarityThreshold = 0.3
): KeywordCluster[] {
  const tokenSets = new Map(keywords.map((kw) => [kw, tokenize(kw)] as const));
  const assigned = new Set<string>();
  const clusters: KeywordCluster[] = [];

  for (const kw of keywords) {
    if (assigned.has(kw)) continue;
    const cluster = [kw];
    assigned.add(kw);
    for (const other of keywords) {
      if (assigned.has(other)) continue;
      if (jaccardSimilarity(tokenSets.get(kw)!, tokenSets.get(other)!) >= similarityThreshold) {
        cluster.push(other);
        assigned.add(other);
      }
    }

    // Name the cluster after the most common significant word across its members.
    const wordCounts = new Map<string, number>();
    for (const member of cluster) {
      for (const word of tokenSets.get(member)!) {
        wordCounts.set(word, (wordCounts.get(word) ?? 0) + 1);
      }
    }
    let clusterName = cluster[0]!;
    let topCount = 0;
    for (const [word, n] of wordCounts) {
      if (n > topCount) {
        topCount = n;
        clusterName = word[0]!.toUpperCase() + word.slice(1);
      }
    }

    clusters.push({ clusterName, keywords: cluster });
  }

  return clusters;
}
